specification/evaluation-context/evaluation-context.json

```diff
@@ -0,0 +1,23 @@
+[
+    {
+        "id": "Requirement 3.1",
+        "clean id": "requirement_3_1",
+        "content": "The `evaluation context` structure MUST define a required `targeting key` field of type string, identifying the subject of the flag evaluation.",
```

Comment:
I think this is in conflict with Hooks 1.1 which says: 
> Hook context MUST provide: the flag key, flag type, evaluation context, and the default value.

If you have to provide an `evaluation context` and an `evaluation context MUST define a required 'targeting key' field`, then that conflicts with Flag Evaluation 1.7 which says context is optional.

> The client MUST provide methods for flag evaluation, with parameters flag key (string, required), default value (boolean | number | string | structure, required), evaluation context (optional), and evaluation options (optional), which returns the flag value.

Previous to this requirement I was specifying an empty EvaluationContext if one didn't exist. Can/should I just generate a random value as the targeting key? Something else?

Reply:
Ya... I understand your confusion. This could be a bit sticky...

Let me try to explain my reasoning for this, and perhaps we can come up with a solution:

Some vendor SDKs require a context parameter (specifically, LaunchDarkly requires a `User`), and that parameter MUST be set for ALL flag evaluations, and MUST have a key (user identifier). Others (Flagsmith) don't require a user identifier, but if there isn't one, attributes cannot be used in the evaluation (the flag evaluation must be "contextless"). 

What I was trying to suggest (and failed to communicate) was that if any context fields were set, the `targetingKey` must also be set, this way, we can at least gracefully handle the Flagsmith case. It doesn't, however, really solve the LaunchDarkly issue. I'm also not sure it's really the right choice. To put all this another way, we have a couple problems:

- How do we deal with vendors that ALWAYS require a user identifier? I don't think we want to make context required... 
- How do we handle SDKs where the vendor doesn't require a user identifier for "contextless" flag evaluations, but requires one if any attributes are defined?

The empty `evaluation context` implementation detail is dependent on what we decide above, which admittedly, I'm a bit stuck on.

cc @beeme1mr

Reply:
@InTheCloudDan Maybe you can weigh in on the above. Is it reasonable to auto-generate a LDUser key? Will that cause inconsistent behavior in percentage rollout situations, for instance?

Reply:
I don't know what the correct behavior is for LD.  I would expect the provider to either say: "You didn't specify a targeting key. Giving you the default value (e.g. light error)" or "You didn't specify a targeting key. I'm going to make one up for you."

Reply:
@justinabrahms we require the users to generate that, our signature enforces the `user/context` to be passed into the SDK at least on the server side.

Reply:
@InTheCloudDan yep, the question is what LD would do for context-less evaluations. All user data is stored in the evaluation context. If a flag evaluation is done without a context parameter being passed, what would a hypothetical launch darkly provider do?

Reply:
@toddbaert we would need to discuss internally, generating an ID is something we have not previously done except maybe in the case of our React SDK on the client side so I would need to confirm that.

Reply:
Just trying to write out ideas, OF could support a `default context` function at the client layer that returns a context if none is passed at an evaluation call site. But it would still be a difference where for some providers it is allowed to be empty and others it cannot be.

Reply:
at Spit.io we have the same problem. 
> Is it reasonable to auto-generate a LDUser key?
At Split, the implications of randomizing this are huge. If a flag is set with a percentage rollout, then a user can get different flag values/treatment as we use consistent hashing for the bucketing logic and rely in the user identity (userId for instance) to be the constant identifier. So without the notion of targeting key, at least Split cannot be a provider you can just add unless the above user code that interacts with the reference SDK changes.

Reply:
What I'm hearing is that either:

-  the targeting key is set through something like a hook
- you have to always pass an evaluation context with at least a targeting key

Anyone see other options?

Reply:
I understand that some providers require it. It's trivially easy to add a global hook to set it, and it will be come even easier to do so once we've implemented implicit context propagation via continuations or thread-local storage, so here's what I propose:

It stays optional. Providers that require it can throw or otherwise error during evaluation with a error log saying it must be set. This will cause them to default, and application authors will see that this provider requires the `targeting key` to be set, and address it with a global hook, or at evaluation. Like I said above, this will become even easier when we can use implicit context propagation to grab user details off of the thread/golang-context/node-continuation.

Thoughts? 

@justinabrahms @patricioe @InTheCloudDan

Reply:
Can there be some more information around setting this as a global hook? So all of the evaluations would have the same key or the hook would return a random key?

For providers that require it this appears to be a suboptimal experience due to the effects it can have on evaluation/return value as @patricioe pointed out.

Reply:
>Can there be some more information around setting this as a global hook? So all of the evaluations would have the same key or the hook would return a random key?

@InTheCloudDan 

Ya I think what I said before was a bit unclear. I meant that a global hook might be used to dynamically add a unique user identifier based on some contextual data. Perhaps you have the user's `email` in the context - you could hash that in a hook and set the `targeting key` that way. At the end of the day, as is the case now with any SDK that requires a "user identifier" of some kind, it will be up to the author to make sure a basis for identifying the user is available at the time of flag evaluation.

The implicit context propagation features that will be spec'd out later will help with this by defining a way to inject values into the flag evaluation dynamically without adding them directly in a flag evaluation, as we do [here](https://github.com/open-feature/playground/blob/main/packages/app/src/app/transaction-context.middleware.ts) in the playground.

I hope that's helpful.